### ISSUE TITLE
object-shorthand rule does not properly handle computed properties

### ISSUE THREAD

**Issue by lemonmade** [eslint/eslint#5871]
**What version of ESLint are you using?**

2.8.0

**What parser (default, Babel-ESLint, etc.) are you using?**

Default

**Please show your full configuration:**

```json
{
  "parserOptions": {
    "ecmaVersion": 6,
    "sourceType": "module"
  },
  "rules": {
    "object-shorthand": ["error", "always"]
  }
}
```

**What did you do? Please include the actual source code causing the issue.**

```js
var foo = {
  bar: function() {},
  ['bar']: function() {},
  [bar]: function() {},

  baz: baz,
  ['baz']: baz
};
```

**What did you expect to happen?**

I would have expected all properties in this object to have errors.

**What actually happened? Please include the actual, raw output from ESLint.**

```
  2:3  error  Expected method shorthand    object-shorthand
  6:3  error  Expected property shorthand  object-shorthand
```

When using computed properties, object shorthand was not checked (admittedly, the string literal computed properties are a pretty dumb case, since they could just be used directly as keys).

I'm happy to take a stab at this one if someone can confirm this is not intentional.

**Comment by nzakas** [comment]
This appears to be working as designed. See comments inline:

```js
var foo = {
  bar: function() {},         // correctly flags this
  ['bar']: function() {},     // does not flag this because it would require interpreting the key
  [bar]: function() {},      // does not flag this because there's no way to know what "bar" is

  baz: baz,                    // correctly flags this
  ['baz']: baz                 // does not flag this because it would require interpreting the key
};
```

So as far as I can tell, this is the correct behavior. Computed properties are intentionally ignored:
https://github.com/eslint/eslint/blob/master/lib/rules/object-shorthand.js#L69

I can see an argument that we should attempt simple computed properties (those that have only a string identifier and nothing else), though I'd want feedback from the rest of the team on that.

**Comment by lemonmade** [comment]
Thanks for responding so quickly! I can see why you wouldn't be able to flag this:

```js
var foo = {
  [bar]: bar,
};
```

Since shorthand properties are predicated on knowing the value of the key. I'm not sure I understand why the method case would be rejected:

```js
var foo = {
  [bar]: function() {},

  // would prefer:
  [bar]() {}
};
```

My understanding was that the shorthand method syntax didn't really have anything to do with the key name itself.

The literals part of this would probably be a reasonable enhancement, even if it is pretty contrived (I've written a `no-useless-computed-properties` rule for my own project to prevent these since there's no real point to them).

**Comment by nzakas** [comment]
For this:

```js
var foo = {
  [bar]: function() {},

  // would prefer:
  [bar]() {}
};
```

You have a good point, I think it makes sense to make that change. Do you want to submit a PR for it?

**Comment by lemonmade** [comment]
Sure thing!
